libs/vkd3d/command.c

```diff
@@ -8345,121 +8345,559 @@ static void STDMETHODCALLTYPE d3d12_command_list_CopyTiles(d3d12_command_list_if
+        {
+            /* Ensure that the required depth and stencil resolve modes are supported */
+            const VkPhysicalDeviceVulkan12Properties *vk12 = &list->device->device_info.vulkan_1_2_properties;
+            VkResolveModeFlagBits vk_mode = vk_resolve_mode_from_d3d12(mode);
+            bool supports_render_pass_resolve = vk12->independentResolveNone;
+
+            for (i = 0; i < region_count && supports_render_pass_resolve; i++)
+            {
+                if (regions[i].dstSubresource.aspectMask & VK_IMAGE_ASPECT_DEPTH_BIT)
+                    supports_render_pass_resolve &= !!(vk12->supportedDepthResolveModes & vk_mode);
+                if (regions[i].dstSubresource.aspectMask & VK_IMAGE_ASPECT_STENCIL_BIT)
+                    supports_render_pass_resolve &= !!(vk12->supportedStencilResolveModes & vk_mode);
+            }
+
+            if (supports_render_pass_resolve)
+            {
+                /* Source and destination regions must be identical for
+                 * resolve attachments. */
+                for (i = 0; i < region_count && supports_render_pass_resolve; i++)
+                {
+                    supports_render_pass_resolve = regions[i].srcOffset.x == regions[i].dstOffset.x &&
+                            regions[i].srcOffset.y == regions[i].dstOffset.y;
+                }
+            }
+
+            if (!supports_render_pass_resolve)
+                path = VKD3D_RESOLVE_PATH_RENDER_PASS_PIPELINE;
+        }
     }
 
-    device = list->device;
-    vk_procs = &device->vk_procs;
-    d3d12_command_list_end_current_render_pass(list, false);
-    d3d12_command_list_end_transfer_batch(list);
+    /* Attachment resolves cannot be used with an offset */
+    if (path == VKD3D_RESOLVE_PATH_RENDER_PASS_ATTACHMENT)
+    {
+        for (i = 0; i < region_count; i++)
+        {
+            if (regions[i].srcOffset.x != regions[i].dstOffset.x ||
+                    regions[i].srcOffset.y != regions[i].dstOffset.y)
+            {
+                path = VKD3D_RESOLVE_PATH_RENDER_PASS_PIPELINE;
+                break;
+            }
+        }
+    }
 
-    if (dst_resource->format->type == VKD3D_FORMAT_TYPE_TYPELESS || src_resource->format->type == VKD3D_FORMAT_TYPE_TYPELESS)
+    /* All alternative code paths require render target usage for the destination image */
+    if (path != VKD3D_RESOLVE_PATH_DIRECT &&
+            !(dst_resource->desc.Flags & (D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET | D3D12_RESOURCE_FLAG_ALLOW_DEPTH_STENCIL)))
     {
-        if (!(vk_format = vkd3d_format_from_d3d12_resource_desc(device, &dst_resource->desc, format)))
+        FIXME("Selected resolve path %u for mode %u, format %u, but destination image cannot be used as a render target.\n",
+                path, mode, format);
+
+        if (vkd3d_format->vk_aspect_mask == VK_IMAGE_ASPECT_COLOR_BIT)
+            path = VKD3D_RESOLVE_PATH_DIRECT;
```

Comment:
This will always lead to incorrect results no matter what, but the only alternative we have would be to allocate a temporary image and then do a copy, which seems painful.

Reply:
I'd imagine we'd return UNSUPPORTED here.

Reply:
Fair, can do that too.